Source/cielim/Private/ProtobufActor.cpp

```diff
@@ -1,35 +1,205 @@
+ */
+FQuat CreateCelestialBodyRotationQuat(const vizProtobufferMessage::VizMessage_CelestialBody& celestialbody) {
+    
+    // Set vectors
+    FVector Rotation1 = FVector4d(celestialbody.rotation(0), celestialbody.rotation(1), celestialbody.rotation(2), 0);
+    FVector Rotation2 = FVector4d(celestialbody.rotation(3), celestialbody.rotation(4), celestialbody.rotation(5), 0);
+    FVector Rotation3 = FVector4d(celestialbody.rotation(6), celestialbody.rotation(7), celestialbody.rotation(8), 0);
+    FVector Rotation4 = FVector4d(0, 0, 0, 1);
+
+    // Set matrix
+    FMatrix Mat = FMatrix(Rotation1, Rotation2, Rotation3, Rotation4);
+
+    // Set Quaternion
+    return FQuat(Mat);
+}
+
+/**
+ * @brief SpawnCelestialBodies() Spawns all celestial bodies from the vizmessage into the level
+ * 
+ */
+void AProtobufActor::SpawnCelestialBodies()
+{
+    if (BpCelestialBody) {
+        for (const auto& celestialbody : vizmessage.celestialbodies()) {
+
+            FVector3d SpawnLocation = FVector3d(celestialbody.position(0), celestialbody.position(1), celestialbody.position(2));
+            FVector3d SpawnLocationLH = Right2LeftVector(SpawnLocation);
+            FQuat Quat = CreateCelestialBodyRotationQuat(celestialbody);
+            FQuat QuatLH = RightQuat2LeftQuat(Quat);
+            FRotator SpawnRotation = FRotator(QuatLH);
+
+            // TODO: Set name
+            FString CbName = FString(celestialbody.bodyname().c_str());
+            // FActorSpawnParameters SpawnParams;
+            // SpawnParams.Name = FName(*CbName);
+
+            ACelestialBody* TempCb = GetWorld()->SpawnActor<ACelestialBody>(BpCelestialBody, SpawnLocationLH, SpawnRotation);
+            TempCb->name = CbName;
+            TempCb->SetRadiusEvent(celestialbody.radiuseq());
+            CelestialBodyArray.Add(TempCb);
+        }
+    }
+    else {
+        UE_LOG(LogTemp, Warning, TEXT("Defualt BpCelestialBody has not been set in BP_ProtobufActor"));
+    }
+}
+
+/**
+ * @brief GetRotatorFromMRP(sigma) Converts an MRP into an Unreal Rotation Container (FRotator) 
+ * 
+ * @param sigma The MRP vector
+ * @return FRotator Unreal Rotaiton Container
+ */
+FRotator GetRotatorFromMRP(FVector3d sigma)
+{
+    FQuat Quat = MRPtoQuaternion(sigma);
+    FQuat QuatLH = RightQuat2LeftQuat(Quat);
+    return FRotator(QuatLH); 
+}
+
+/**
+ * @brief Basilisk2UnrealCamera(sigma_CB)
+ * 
+ * @param sigma_CB 
+ * @return FVector3d 
+ */
+FVector3d Basilisk2UnrealCamera(FVector3d sigma_CB)
+{
+    FQuat q = MRPtoQuaternion(sigma_CB); 
+    FQuat q_temp = MRPtoQuaternion(FVector3d(-1./3, -1./3, 1./3)); // Temporary correction for Basilisk output CrCu (r unreal u basilisk)
+    FQuat q_correction = FQuat(0.5, -0.5, 0.5, 0.5);
+    return QuaterniontoMRP(q * q_correction * q_temp);
+}
+
+/**
+ * @brief SpawnSpacecraft() Spawns all spacecraft from the vizmessage into the level
+ * 
+ */
+void AProtobufActor::SpawnSpacecraft()
+{
+    if (BpSpacecraft) {
```

Comment:
:small_red_triangle_down: We can invert this logic to make the function read simpler. Check if not blue print, emit error and return, no else condition, just continue on with the function.

Reply:
Fixed and did the same logic for spawning celestial bodies